drift_dev/lib/src/model/column.dart

```diff
@@ -158,7 +160,7 @@ class MoorColumn implements HasDeclaration, HasType {
 
     // We currently use nullable columns everywhere because it's not clear how
     // to express nullability in joins otherwise.
-    return options.nnbd ? '$code?' : code;
+    return options.nnbd && (!checkNullable || nullable) ? '$code?' : code;
```

Comment:
The comment above is kind of confusing now, let's add a quick note about how it needs needs to be non-null to pass the right type to the serializer.

Reply:
Updated to:
```
    // We currently use nullable columns everywhere because it's not clear how
    // to express nullability in joins otherwise. When serializing data with
    // JsonTypeConverter it needs to match the nullability of the column
```